### PR TITLE
teleinfo: use text_sensor and sensor.

### PR DESCRIPTION
**Pull request by 0hax**
## Description:

Create a Teleinfo platform and add possibility to define sensor or
text_sensor separately.
Without this commit there was no way to read a Teleinfo label returning
text instead of numbers.
Now that text_sensor is supported, once can publish state about 'text'
information too.

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#871

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/teleinfo/teleinfo.h [hunk 1]
```diff
@@ -14,20 +13,21 @@ static const uint8_t MAX_TAG_SIZE = 64;
...
 static const uint16_t MAX_BUF_SIZE = 1024;
 
-struct TeleinfoSensorElement {
-  const char *tag;
-  sensor::Sensor *sensor;
+class TeleInfoListener {
+ public:
+  std::string tag;
+  virtual void publish_val(std::string val){};
+  virtual void dump_config(){};
```
Comment (jesserockz) [inline esphome/components/teleinfo/teleinfo.h:20]:
```suggestion
  virtual void publish_val(std::string val);
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
Just one small change i came across

File: esphome/components/teleinfo/teleinfo.h (hunk 1 above)
Comment (0hax) [reply esphome/components/teleinfo/teleinfo.h:20]: Don't we need dump_config() in `TeleInfoListener` class as it is used in `TeleInfo::dump_config()` via `element->dump_config();`
The function is defined in `esphome/components/teleinfo/sensor/teleinfo_sensor.cpp` for example.

Comment (jesserockz) [reply esphome/components/teleinfo/teleinfo.h:20]: `TeleInfoListener` does not need it. `Component` defines it for all components. 
You don't need to `dump_config` manually for any child sensor as the method will be called by the core code. This way would mean each sensor is dumped twice.

Comment (0hax) [reply esphome/components/teleinfo/teleinfo.h:20]: Ah right, thanks for explaining.

**Comment by 0hax** [comment]
@jesserockz I have applied your suggestion and rebased to latest dev.

**Review by glmnet — APPROVED** [review]

**Comment by 0hax** [comment]
Should I fix the conflict or you are going to do that?

**Comment by glmnet** [comment]
Ahh please keep it ready to merge

**Comment by 0hax** [comment]
> Ahh please keep it ready to merge

Ok I will rebase it

**Review by jesserockz — APPROVED** [review]

**Comment by jesserockz** [comment]
@glmnet Considering this is a `breaking change`, do you want it in `1.16.0` as the original code is only there too so if this was merged and released, then its not breaking to stable, only within dev/beta

**Comment by 0hax** [comment]
Did you merge this somewhere at the end ?
It would be good to have it in the final release with the update on the doc as well: #871

**Comment by 0hax** [comment]
Any news on this guys? @jesserockz @glmnet 
I think it would be good to merge this

**Comment by 0hax** [comment]
Hi @jesserockz @glmnet
Do you need anything from me before merging this ?

**Comment by jesserockz** [comment]
> Hi @jesserockz @glmnet
> Do you need anything from me before merging this ?

Merge conflicts by the looks 😉
But otherwise I think I was happy. Will give a quick look over after the conflicts are fixed

**Comment by 0hax** [comment]
> > Hi @jesserockz @glmnet
> > Do you need anything from me before merging this ?
> 
> Merge conflicts by the looks wink
> But otherwise I think I was happy. Will give a quick look over after the conflicts are fixed

Ahah sorry I was thinking you were not sure about merging it.
I have fixed the merge conflicts, not sure why but the lint test is failing during the test whereas it's ok locally.
Known issue?
Running `black sphome/components/teleinfo/__init__.py` is fine locally

**Comment by jesserockz** [comment]
It's the sensor and text_sensor init files not the component init file that need black.

**Comment by 0hax** [comment]
> It's the sensor and text_sensor init files not the component init file that need black.

Ah right, should be fine now. Thank you

**Comment by damienclaveau** [comment]
@0hax I have deployed the dev branch with your evolution today, and it's working like a charm.
Thank you for your work on the teleinfo component.

and thank you jesserocks for this awesome project.

**Comment by 0hax** [comment]
> @0hax I have deployed the dev branch with your evolution today, and it's working like a charm.
> Thank you for your work on the teleinfo component.
> 
> and thank you jesserocks for this awesome project.

Thanks for your feedback, I'm happy to hear it's useful not only for me :)

**Pull request merged**